src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -33,6 +35,42 @@ internal sealed class PoolingDataSource : SqlDataSource
         private static int _objectTypeCount; // EventSource counter
         private readonly int _objectID = Interlocked.Increment(ref _objectTypeCount);
 
+        //TODO: readonly TimeSpan _connectionLifetime;
+
+        // Counts the total number of open connectors tracked by the pool.
+        volatile int _numConnectors;
+
+        // Counts the number of connectors currently sitting idle in the pool.
+        volatile int _idleCount;
+
+        // Prevents synchronous operations from blocking on all available threads,
+        // which would stop async tasks from being scheduled and cause deadlocks.
+        private static SemaphoreSlim SyncOverAsyncSemaphore { get; } = new(Math.Max(1, Environment.ProcessorCount / 2));
+
+        /// <summary>
+        /// Tracks all connectors currently managed by this pool, whether idle or busy.
+        /// Only updated rarely - when physical connections are opened/closed - but is read in perf-sensitive contexts.
+        /// </summary>
+        private readonly SqlConnector?[] _connectors;
+
+
+        /// <summary>
+        /// Reader side for the idle connector channel. Contains nulls in order to release waiting attempts after
+        /// a connector has been physically closed/broken.
+        /// </summary>
+        readonly ChannelReader<SqlConnector?> _idleConnectorReader;
+        internal ChannelWriter<SqlConnector?> IdleConnectorWriter { get; }
+
+        internal sealed override (int Total, int Idle, int Busy) Statistics
```

Comment:
I agree, and it also broaches the topic of OTel metrics support (#2211.) We need to be able to summarise multiple PoolingDataSource statistics into a single set of "per-connection-string" metric values, so these statistics are needed.

Small nit: I'd prefer to read both the total and idle connection figures atomically, since they're so closely related. If both fields are in a class, we can use `FieldOffset` to position them at 0 and 4 bytes, have a `ulong` positioned at 0 bytes, then use `Interlocked.Read` on the `ulong` to perform an atomic read of both values at once.

Reply:
Let me capture this in a new issue. I think this deserves focus but I don't want it to block this PR at the moment.

Reply:
#2708